src/install.rs

```diff
@@ -24,6 +24,65 @@ use crate::download::*;
 use crate::errors::*;
 use crate::source::*;
 
+/// Integrity verification hash for an Ignition config.
+#[derive(Debug)]
+pub enum IgnitionHash {
+    /// SHA-512 digest.
+    Sha512(Vec<u8>),
+}
+
+impl IgnitionHash {
+    /// Try to parse an hash-digest argument.
+    ///
+    /// This expects an input value following the `ignition.config.verification.hash`
+    /// spec, i.e. `<type>-<value>` format.
+    pub fn try_parse(input: &str) -> Result<Self> {
+        let parts: Vec<_> = input.splitn(2, '-').collect();
+        if parts.len() != 2 {
+            bail!("failed to detect hash-type and digest in '{}'", input);
+        }
+        let (hash_kind, hex_digest) = (parts[0], parts[1]);
+
+        let hash = match hash_kind {
+            "sha512" => {
+                let digest = hex::decode(hex_digest).chain_err(|| "decoding hex digest")?;
+                ensure!(
+                    digest.len().saturating_mul(8) == 512,
+                    "wrong digest length ({})",
+                    digest.len().saturating_mul(8)
+                );
+                IgnitionHash::Sha512(digest)
+            }
+            x => bail!("unknown hash type '{}'", x),
+        };
+
+        Ok(hash)
+    }
+
+    /// Digest and validate input data.
+    pub fn validate<R>(&self, input: &mut R) -> Result<()>
+    where
+        R: Read,
+    {
+        use sha2::digest::Digest;
+
+        let (mut hasher, digest) = match self {
+            IgnitionHash::Sha512(val) => (sha2::Sha512::new(), val),
+        };
+        let _n = copy(input, &mut hasher).chain_err(|| "copying input to hasher")?;
+        let computed = hasher.result();
+
+        if computed.as_slice() != digest.as_slice() {
+            bail!(
+                "failed to validate hash, mismatch with computed '{}'",
```

Comment:
I didn't do that because it will make the error message very long, and the expected hash should be coming from the user at a calling site very close to this.
That said, I am open to add it. Do you still prefer to have both logged?

Reply:
In those types of "didn't get what I expected" error messages, I tend to lean towards just providing both the actual vs expected to try to be helpful.

E.g. here for example, having the actual digest found means that once you verify that it's indeed correct and fix your `--ignition-hash` argument, you're pretty confident it'll work.